src/drivers/imx/esai.c

```diff
@@ -3,38 +3,410 @@
+	dai_write(dai, REG_ESAI_TSMA, 0);
+	dai_write(dai, REG_ESAI_TSMB, 0);
+	dai_write(dai, REG_ESAI_RSMA, 0);
+	dai_write(dai, REG_ESAI_RSMB, 0);
+
+	/* Program FIFOs -- rx only reset them
+	 * TODO: When we will support recording this needs to be adjusted.
+	 */
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, 0);
+
+	/* Reset transmit FIFO */
+	dai_update_bits(dai, REG_ESAI_TFCR,
+			ESAI_xFCR_xFR_MASK,
+			ESAI_xFCR_xFR);
+
+	/* Set transmit fifo configuration register
+	 * xWA(24): 24-bit samples as input. Must agree with xSWS above.
+	 *	    TODO get sample width from topology.
+	 * xFWM(96): Trigger next DMA transfer when at least 96 (of the 128)
+	 *	     slots are empty.
+	 * TE(1): Enable 1 transmitter.
+	 * TIEN: Transmitter initialization enable. This will pull the initial
+	 *       samples from the FIFO in the transmit registers. The
+	 *       alternative would have been to manually initialize the
+	 *       transmit registers manually, which would have been more
+	 *       complex to implement.
+	 */
+	dai_update_bits(dai, REG_ESAI_TFCR,
+			ESAI_xFCR_xFR_MASK | ESAI_xFCR_xWA_MASK
+			| ESAI_xFCR_xFWM_MASK | ESAI_xFCR_TE_MASK
+			| ESAI_xFCR_TIEN,
+			ESAI_xFCR_xWA(24) | ESAI_xFCR_xFWM(96)
+			| ESAI_xFCR_TE(1) | ESAI_xFCR_TIEN);
+
+	/* Set the clock divider to divide EXTAL by 16 (DIV8 from PSR,
+	 * plus a divide by 2 which is mandatory overall)
+	 * This configuration supports hardcoded MCLK at 49152000 Hz and
+	 * obtains frame clock of 96000 Hz (2 48000Hz channels) and bit clock
+	 * of 3072000 Hz (32-bit samples).
+	 * xFP(1): No division from this divider (can do 1-16)
+	 * xPSR_DIV8: Divide by 8
+	 * There is also an additional divide by 2 which is forced by the
+	 * hardware design of the ESAI.
+	 *
+	 * TODO use ESAI params instead of hardcode to compute the clock divider
+	 * settings.
+	 */
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xFP_MASK,
+			ESAI_xCCR_xFP(1));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xPSR_MASK,
+			ESAI_xCCR_xPSR_DIV8);
+
+	/* Remove ESAI personal reset */
+	dai_update_bits(dai, REG_ESAI_xCR(0), ESAI_xCR_xPR, 0);
+	dai_update_bits(dai, REG_ESAI_xCR(1), ESAI_xCR_xPR, 0);
+
+	/* Configure ESAI pins, enable them all */
+	dai_update_bits(dai, REG_ESAI_PRRC, ESAI_PRRC_PDC_MASK,
+			ESAI_PRRC_PDC(ESAI_GPIO));
+	dai_update_bits(dai, REG_ESAI_PCRC, ESAI_PCRC_PC_MASK,
+			ESAI_PCRC_PC(ESAI_GPIO));
 	return 0;
 }
 
+static void esai_start(struct dai *dai, int direction)
+{
+	/* FIFO enable */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFEN_MASK,
+			ESAI_xFCR_xFEN);
+
+	/* To prevent channel swap issues, write a sample per channel. The ESAI
+	 * can begin transmitting these samples while it is requesting for a
+	 * DMA transfer, before said DMA transfer actually puts more samples
+	 * into the ESAI FIFO. Only needed on playback/transmit.
+	 */
+	if (!direction) {
+		dai_write(dai, REG_ESAI_ETDR, 0);
```

Comment:
I think it might be worth adding a TODO comment to remind us that we should at least write a word per active channel

Reply:
Done